modules/wallet/transactionbuilder.go

```diff
@@ -9,6 +9,90 @@ import (
 	"github.com/NebulousLabs/Sia/types"
 )
 
+// walletSpender allows clients to retrieve and modify keys and outputs in a
+// wallet in order to spend siacoins.
+type walletSpender interface {
+	// acquireLock acquires a mutex on walletSpender. Clients must call
+	// acquireLock before calling any other function in walletSpender and must
+	// call releaseLock before exiting scope.
+	acquireLock()
+	releaseLock()
+
+	siacoinOutputs() map[types.SiacoinOutputID]types.SiacoinOutput
+	siafundOutputs() map[types.SiafundOutputID]types.SiafundOutput
+	unconfirmedProcessedTransactions() []modules.ProcessedTransaction
+
+	// spendHeight returns the block height at which a given output was spent.
+	spendHeight(types.OutputID) types.BlockHeight
+
+	// setSpendHeight marks the output with the given ID as spent in the given
+	// block.
+	setSpendHeight(types.OutputID, types.BlockHeight)
+
+	// key returns the key associated with the given hash.
+	key(types.UnlockHash) spendableKey
+
+	// hasKey indicates whether a key with the given hash exists in the wallet.
+	hasKey(types.UnlockHash) bool
+
+	// consensusSetHeight returns the current block height of the wallet.
+	consensusSetHeight() types.BlockHeight
+
+	// nextPrimarySeedAddress fetches the next address from the primary seed.
+	nextPrimarySeedAddress() (types.UnlockConditions, error)
+}
```

Comment:
This doesn't really succeed at the goal of decoupling the transaction builder from the wallet. All you really did was wrap the used fields with setters and getters to the wallet, and not in a way that promotes type safety.

As much as possible, you should avoid getters and setters, and instead focus on the bigger picture. What you really need is some way to select a set of outputs to fund a transaction, to get signatures for those outputs, and then to have some control over indicating whether they are 'reserved' or not (that's all the 'spendHeight' is for - it tells the wallet not to use that output again for a while).

I think you can reduce this interface to about 4 functions:

`findOutputs`, `signTransaction` (implementation details on this one might be tricky), `reserveOutputs`, `releaseOutputs`.

But you definitely don't want to be exposing the inner state of the wallet. It defeats the point of decoupling and adds a lot of overhead when trying to figure out what's safe to modify and whatnot.